### DISCUSSION TITLE
kabelsalat: a live coding version of noisecraft

### DISCUSSION THREAD

**Discussion by felixroos** [maximecb/noisecraft#109] [Show and tell]
Hello there, I thought I'd share my tool [kabelsalat](https://github.com/felixroos/kabelsalat) (think cable salad), which is very much based on the noisecraft audio engine. Instead of a patching UI, it uses code as input, here's a simple example:

```js
sine(220).mul(sine(2).range(.5, 1)).out()
```

That code will generate these nodes:

<img width="405" alt="image" src="https://github.com/maximecb/noisecraft/assets/12023032/102cbcd9-d101-4f23-b9a1-11ab155a6082">

listen to it [here](https://felixroos.github.io/kabelsalat/#c2luZSgyMjApLm11bChzaW5lKDIpLnJhbmdlKC41LCAxKSkub3V0KCk=) or, for a more involved example check out [the acid machine example with drums](https://felixroos.github.io/kabelsalat/#Ly8gbWF4aW1lY2IgLSB0aGUgbGl0dGxlIGFjaWQgbWFjaGluZSB0aGF0IGNvdWxkCi8vICsgbWF4aW1lY2IgZHJ1bSBtYWNoaW5lCi8vIGNvZGUgYnkgZnJvb3MKLy8gbm9pc2VjcmFmdCAtPiBrYWJlbHNhbGF0Ci8vIGh0dHBzOi8vbm9pc2VjcmFmdC5hcHAvNDcKCmxldCBraWNrID0gZ2F0ZSA9PiBnYXRlLmFkc3IoMCwuMTEsMCwuMTEpCiAgLmFwcGx5KGVudiA9PiBlbnYubXVsKGVudikKICAgIC5tdWwoMTU4KSAvLyBmcmVxdWVuY3kgcmFuZ2UKICAgIC5zaW5lKGVudikKICAgIC5kaXN0b3J0KC44NSkKICApCgpsZXQgc25hcmUgPSBnYXRlID0+IGdhdGUuYWRzcigwLC4xMSwuMSwuMSkKICAubXVsKG5vaXNlKCkpLmZpbHRlciguNzgsLjI5KQoKbGV0IGMgPSBjbG9jaygxNjApCmxldCBub3RlcyA9IGMuY2xvY2tkaXYoOCkKLnNlcSgyNywyNywzOSw1MSwwLDAsMjcsMjcsNDIsMjcsNDAsMCwzMSwzMSw1Niw1MSkKCmxldCBlbnYgPSBub3Rlcy5hZHNyKDAsLjMsLjM0LC41OSkKCm5vdGVzCi5hcHBseTIoaG9sZCkgLy8gaG9sZCBmcmVxIGFib3ZlIDBzCi5taWRpbm90ZSgpCi5zbGlkZSgKICBzaW5lKC4yMSkucmFuZ2UoMCwxKQopCi5wdWxzZSguNDgpCi5tdWwoZW52KQouZmlsdGVyKAogIGVudi5tdWwoc2luZSguMDkpLnJhbmdlKC41NSwxKSksIC8vIGN1dG9mZgogIHNpbmUoLjIyKS5yYW5nZSgwLC4zNSkgLy8gcmVzCikKLmRpc3RvcnQoCiAgc2luZSguMTgpLnJhbmdlKDAsLjg1KQopCi5tdWwoMC41KQoubXVsKGMuY2xvY2tkaXYoMTYpLnNlcSguMjUsMSkpIC8vIHNpZGVjaGFpbgouYWRkKGMuY2xvY2tkaXYoMzIpLnNlcSgxLDEpLmFwcGx5KGtpY2spKQouYWRkKGMuY2xvY2tkaXYoMzIpLnNlcSgwLDEpLmFwcGx5KHNuYXJlKSkKLm91dCgpCg==)

here's an overview of what I've used and adapted:

- most of audiograph.js , but without the concept of params + with a simpler sequencer + some other tweaks
- audioview.js without ui related logic
- synth.js + worklet.js
- I've written a new compiler for learning purposes + to add a generic feedback node

My goal with this is to implement a new audio engine for my other project [strudel](https://strudel.cc/), which currently uses regular web audio nodes, which is kind of limiting and also less performant.

I hope it is ok that I took a lot from this project and repurposed it, but I think this is what free software is about? 
Thank you so much @maximecb for creating noisecraft and also making sure the codebase is so elegant and non-bloated! 
The mechanism of transforming a graph data structure into a super efficient piece of jit optimizable code blows my mind.

Have a nice day,
Felix

**Comment by maximecb** [comment]
Hi felix! It's cool to see that you've built something with the noisecraft audio engine.

Yes it's OK to reuse the code. However, in theory you should be including the [original open source license](https://github.com/maximecb/noisecraft/blob/main/LICENSE) for the files/code that you are using. It would also be nice to include a note in the readme to say that you based your audio synthesis engine on noisecraft and/or that the project was inspired by noisecraft, if you don't mind. That being said, I am very glad to see you found the code useful and that it is working well for you :)

> Thank you so much @maximecb for creating noisecraft and also making sure the codebase is so elegant and non-bloated!

I just can't stand codebases with lots of dependencies 😅 

> [strudel](https://strudel.cc/), which currently uses regular web audio nodes, which is kind of limiting and also less performant.

Are you saying that web audio noses perform less well than noisecraft's audio synthesis engine? If so, that is very surprising considering that noisecraft's audio rendering loop is single threaded and relies on compilation to JavaScript. I wouldn't know though because I never used anything except's web audio's script and output nodes.

I originally created this audio synthesis engine because I felt like the web audio API was just too big/bloated. IMO it's just unlikely that different browser implementations are going to sound exactly the same... I also felt like there was no reason that API needed to be so complex, all we really need is audio input and output!
